include/triton/Dialect/Triton/IR/TritonOps.td

```diff
@@ -438,4 +467,45 @@ def TT_AssertOp : TT_Op<"assert", [MemoryEffects<[MemWrite]>]> {
   let assemblyFormat = "$condition `,` $message `,` $file `,` $func `,` $line attr-dict `:` type($condition)";
 }
 
+//
+// Make a Tile Pointer
+//
+def TT_MakeTilePtrOp : TT_Op<"make_tile_ptr",
+                             [Pure,
+                              SameVariadicOperandSize,
+                              TypesMatchWith<"infer pointer type from the result type",
+                                             "result", "base",
+                                             "getPointerType(getElementTypeOfTilePointerType($_self))">]> {
+  let summary = "Make a tile pointer type with meta information of the parent tensor and the tile";
+
+  let description = [{
+      `tt.make_tile_ptr` takes both meta information of the parent tensor and the tile and returns a pointer to a tile,
+      e.g. returns a type of `tt.ptr<tensor<8x8xf16>>`.
+  }];
+
+  let arguments = (ins
+    TT_Ptr:$base,
+    Variadic<I64>:$shape,
+    Variadic<I64>:$strides,
+    Variadic<I32>:$offsets,
+    DenseI32ArrayAttr:$order
+  );
+
+  let results = (outs TT_Ptr:$result);
```

Comment:
What's TT_Ptr here? The result should have TilePointer type

Reply:
Fixed. To do this, I refactored `TT_Ptr` with another `TT_PtrOf<>`. Now the type system related to pointers is like:
```
// Scalar Pointer Type: `ptr<>`
def TT_Ptr : TT_PtrOf<[AnyType]>;

// Tensor of Pointer Type
def TT_PtrTensor : TensorOf<[TT_Ptr]>;

// Tensor of Pointer Type or Pointer type: `tensor<ptr<>>` or `ptr<>`
def TT_PtrLike : AnyTypeOf<[TT_Ptr, TT_PtrTensor]>;

// Tensor Type
def TT_FpIntTensor : AnyTypeOf<[TT_FloatTensor, TT_IntTensor]>;
def TT_Tensor : AnyTypeOf<[TT_FpIntTensor, TT_PtrTensor]>;

// Pointer Type to Tensor Type: `ptr<tensor<>>`
def TT_TensorPtr : TT_PtrOf<[TT_Tensor]>;

// Any Type in Triton IR
def TT_Type : AnyTypeOf<[TT_FloatLike, TT_IntLike, TT_PtrLike, TT_TensorPtr]>;
```

Reply:
Sorry, but I find the whole terminology very confusing:
```
// Scalar Pointer Type: `ptr<>`
def TT_Ptr : TT_PtrOf<[AnyType]>;
``` 
But `AnyType` can be a Tensor right? So the invariant here isn't enforced. I also don't see the purpose of `PtrOf` ?  The PointeeType information seems to be already present here https://github.com/openai/triton/blob/556ca5504bef702d8f27d7885b4b59db31b19576/include/triton/Dialect/Triton/IR/TritonTypes.td#L58. I think the addition of `PtrOf` creates a lot of confusing redundancy. I feel like there has to be an optrait that just let us constraint the pointeeType of the original `TT_Ptr` here?

Reply:
@ptillet 

Yes, `AnyType` can be a `Tensor`, but we can change `let results = (outs TT_Ptr:$result);` into `let results = (outs TT_TensorPtr:$result)` (where `def TT_TensorPtr : TT_PtrOf<[TT_Tensor]>`)

I felt it is kind of redundancy too at the beginning of implementing this, but I later refer to `RankedTensorType` and `TensorOf` in MLIR and don't think it's redundant anymore. They did the same thing as here, `TensorOf` class is more likely to be a type in the TableGen system while `RankedTensorType` exists in C++ code. And `TensorOf` generates code with a `cppClassName` to do checks.

The whole structure now is:
```
// Pointer Type in TableGen
// The code is copied & modified from the parent class of `TensorOf`: https://github.com/llvm/llvm-project/blob/943df86c82b1450bde45678757b7c39e459bc6ad/mlir/include/mlir/IR/OpBase.td#L536
class TT_PtrOf<list<Type> pointeeTypes> :
    DialectType<Triton_Dialect,
                And<[CPred<"$_self.isa<::mlir::triton::PointerType>()">,
                     Concat<"[](::mlir::Type pointeeType) { return ",
                            SubstLeaves<"$_self", "pointeeType", AnyTypeOf<pointeeTypes>.predicate>,
                                        "; }($_self.cast<::mlir::triton::PointerType>().getPointeeType())">]>,
                "ptr", "::mlir::triton::PointerType">;

// Pointer Type in C++ (corresponding to `TT_PtrOf`)
def TT_PtrType : TritonTypeDef<"Pointer", "ptr"> {
    let summary = "Pointer type (`::mlir::triton::PointerType`) in Triton IR type system";
}
```

After code generation of TableGen, `TT_PtrOf` no longer exists. Only a class called `mlir::triton::PointerType` will exist in `Types.h.inc`:
```
namespace mlir {
class AsmParser;
class AsmPrinter;
} // namespace mlir
namespace mlir {
namespace triton {
class PointerType;
namespace detail {
struct PointerTypeStorage;
} // namespace detail
class PointerType : public ::mlir::Type::TypeBase<PointerType, ::mlir::Type, detail::PointerTypeStorage> {
public:
  using Base::Base;
  static PointerType get(Type pointeeType, int addressSpace);
  static constexpr ::llvm::StringLiteral getMnemonic() {
    return {"ptr"};
  }

  static ::mlir::Type parse(::mlir::AsmParser &odsParser);
  void print(::mlir::AsmPrinter &odsPrinter) const;
  Type getPointeeType() const;
  int getAddressSpace() const;
};
} // namespace triton
} // namespace mlir
MLIR_DECLARE_EXPLICIT_TYPE_ID(::mlir::triton::PointerType)

#endif  // GET_TYPEDEF_CLASSES
```

But in generated `Ops.cpp.inc`, there will be another type check generated by `TT_PtrOf` in the results of `make_tensor_ptr`:
```
  for (::mlir::Type type : resultTypes) {
    (void)type;
    if (!(((type.isa<::mlir::triton::PointerType>())) && ([](::mlir::Type pointeeType) { return ((((pointeeType.isa<::mlir::TensorType>())) && ([](::mlir::Type elementType) { return ((elementType.isFloat8E4M3FN())) || ((elementType.isFloat8E5M2())) || ((elementType.isF16())) || ((elementType.isBF16())) || ((elementType.isF32())) || ((elementType.isF64())); }(pointeeType.cast<::mlir::ShapedType>().getElementType()))) || (((pointeeType.isa<::mlir::TensorType>())) && ([](::mlir::Type elementType) { return ((elementType.isSignlessInteger(1))) || ((elementType.isSignlessInteger(8))) || ((elementType.isSignlessInteger(16))) || ((elementType.isSignlessInteger(32))) || ((elementType.isSignlessInteger(64))); }(pointeeType.cast<::mlir::ShapedType>().getElementType())))) || (((pointeeType.isa<::mlir::TensorType>())) && ([](::mlir::Type elementType) { return ((elementType.isa<::mlir::triton::PointerType>())) && ([](::mlir::Type pointeeType) { return (true); }(elementType.cast<::mlir::triton::PointerType>().getPointeeType())); }(pointeeType.cast<::mlir::ShapedType>().getElementType()))); }(type.cast<::mlir::triton::PointerType>().getPointeeType())))) {
      return parser.emitError(parser.getNameLoc()) << "'result' must be ptr, but got " << type;
    }
  }
```

I was just curious about how MLIR does `TensorOf` and `RankedTensorType` and learned it. And I think it's a more general and "MLIR style" solution.

If you still think it is confusing, I can change to the `OpTrait` way, actually that's much easier & fine for me :)

Reply:
With this, we can mark the `ptr` in `tt.load/store` as:
```
let arguments = (ins AnyTypeOf<[TT_PtrLike, TT_TensorPtr]>:$ptr, TT_Type:$value, Optional<TT_BoolLike>:$mask,
                         OptionalAttr<DenseI32ArrayAttr>:$boundaryCheck,
                         DefaultValuedAttr<TT_CacheModifierAttr, "triton::CacheModifier::NONE">:$cache,
                         DefaultValuedAttr<TT_EvictionPolicyAttr, "triton::EvictionPolicy::NORMAL">:$evict);
```

Reply:
I see. Yes, that makes sense. I think I agree with this, but my only comment is that `def TT_Ptr : TT_PtrOf<[AnyType]>;` seems essentially identical to `TT_PtrLike` unless I'm missing something?

Reply:
Never mind

Reply:
Yeah, I think it's fine! Thanks for the explanation.